src/qt/trafficgraphwidget.cpp

```diff
@@ -1,257 +1,305 @@
+{
+    int h = height() - YMARGIN * 2;
+    int w = width() - XMARGIN * 2;
+    int lastY = 0;
+
+    path.moveTo(XMARGIN, YMARGIN + h);
+
+    int samplesIndex = samplesHead;
+    for (int i = 0; i < samples.size(); ++i) {
+        int x = w * i / NUM_DISPLAYED_SAMPLES;
+        int y = h - (int)(h * samples[samplesIndex] / fMax);
+
+        path.lineTo(XMARGIN + x, YMARGIN + y);
+
+        samplesIndex++;
+        if (samplesIndex >= samples.size()) {
+            samplesIndex = 0;
+        }
+
+        lastY = y;
+    }
+    path.lineTo(XMARGIN + w, YMARGIN + lastY);
+    path.lineTo(XMARGIN + w, YMARGIN + h);
+}
+
+void TrafficGraphWidget::paintEvent(QPaintEvent *)
+{
+    QPainter painter(this);
+    painter.fillRect(rect(), Qt::black);
+
+    if (fMax <= 0.0f) {
+        return;
+    }
+
+    QColor axisCol(Qt::gray);
+    int h = height() - YMARGIN * 2;
+    painter.setPen(axisCol);
+    painter.drawLine(XMARGIN, YMARGIN + h, width() - XMARGIN, YMARGIN + h);
+
+    // decide what order of magnitude we are
+    int base = floor(log10(fMax));
+    float val = pow(10.0f, base);
+
+    const QString units     = tr("KB/s");
+    const float yMarginText = 2.0;
+
+    // draw lines
+    painter.setPen(axisCol);
+    painter.drawText(XMARGIN, YMARGIN + h - h * val / fMax-yMarginText, QString("%1 %2").arg(val).arg(units));
+    for(float y = val; y < fMax; y += val) {
+        int yy = YMARGIN + h - h * y / fMax;
+        painter.drawLine(XMARGIN, yy, width() - XMARGIN, yy);
+    }
+
+    // if we drew 3 or fewer lines, break them up at the next lower order of magnitude
+    if(fMax / val <= 3.0f) {
+        axisCol = axisCol.darker();
+        val = pow(10.0f, base - 1);
+        painter.setPen(axisCol);
+        painter.drawText(XMARGIN, YMARGIN + h - h * val / fMax-yMarginText, QString("%1 %2").arg(val).arg(units));
+        int count = 1;
+        for(float y = val; y < fMax; y += val, count++) {
+            // don't overwrite lines drawn above
+            if(count % 10 == 0)
+                continue;
+            int yy = YMARGIN + h - h * y / fMax;
+            painter.drawLine(XMARGIN, yy, width() - XMARGIN, yy);
+        }
+    }
+
+    QPainterPath samplesPathIn;
+    paintPath(samplesPathIn, vTimespanSamplesIn, nTimespanHeadIndex);
+    painter.fillPath(samplesPathIn, QColor(0, 255, 0, 128));
+    painter.setPen(Qt::green);
+    painter.drawPath(samplesPathIn);
+
+    QPainterPath samplesPathOut;
+    paintPath(samplesPathOut, vTimespanSamplesOut, nTimespanHeadIndex);
+    painter.fillPath(samplesPathOut, QColor(255, 0, 0, 128));
+    painter.setPen(Qt::red);
+    painter.drawPath(samplesPathOut);
+}
+
+void TrafficGraphWidget::updateTimespanRates()
+{
+    if (!clientModel) {
+        return;
+    }
+
+    int64_t currentTimeMs = GetTimeMillis();
+    int64_t elapsedMs = currentTimeMs - nTimespanLastSampleTimeMs;
+    if (elapsedMs <= 0) {
+        return;
+    }

```

Comment:
I appreciate that we won't have a visual display bug when returning to standard time from daylight saving time!

Reply:
It wants me to say something in order to resolve this. :) Hopefully yes a differential in ms should always be agnostic to human date weirdness.